### ISSUE TITLE
panic: runtime error: index out of range [-1]

### ISSUE THREAD

**Issue by laur89** [authelia/authelia#2349]
### Description

Crash at a random point; no certain way for reproducing it.
Service has been stable for ages, looks like it's a one-off. Possible notable bit is the issue occurred after [drone](https://www.drone.io/), running on the same docker host, was added to the environment. Relevant bypass rule in Auhtelia:
```
  networks:
    - name: internal
      networks:
        - 172.19.0.0/16
        - 172.17.0.0/16

  rules:
    - domain:
        - mydomain.com
        - "*.mydomain.com"
      policy: bypass
      networks:
        - internal
```

### Expected Behaviour

_N/A_

### Reproduction Steps

_N/A_

### Additional Information

Docker image id: `564d45bfafad` (tracking latest; image inspect at the bottom)
Logs at the time of service crash:

```
time="2021-08-27T10:45:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:45:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:46:27+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:46:27+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:46:27+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:46:27+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:46:27+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:46:27+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:46:27+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:46:27+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:46:27+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:46:27+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:07+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
time="2021-08-27T10:47:47+02:00" level=debug msg="Check authorization of subject username= groups= ip=172.19.0.1 and object https://drone.mydomain.com/rpc/v2/stage (method )."
panic: runtime error: index out of range [-1]

goroutine 15040 [running]:
math/rand.(*rngSource).Uint64(...)
math/rand/rng.go:249
math/rand.(*rngSource).Int63(0xc000121500, 0x4917bffa6fbf717b)
math/rand/rng.go:234 +0x98
github.com/savsgio/gotils/bytes.Rand(0xc0000c6780, 0x20, 0x20, 0xc0000c6780, 0x0, 0x0)
github.com/savsgio/gotils@v0.0.0-20210617111740-97865ed5a873/bytes/bytes.go:19 +0xad
github.com/fasthttp/session/v2.(*Config).defaultSessionIDGenerator(...)
github.com/fasthttp/session/v2@v2.4.1/config.go:33
github.com/fasthttp/session/v2.(*Session).Get(0xc00051a000, 0xc0004b8580, 0x1275f00, 0x7, 0x9cbbcb6400000000)
github.com/fasthttp/session/v2@v2.4.1/session.go:156 +0x204
github.com/authelia/authelia/v4/internal/session.(*Provider).GetSession(0xc00041c450, 0xc0004b8580, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
github.com/authelia/authelia/v4/internal/session/provider.go:78 +0x75
github.com/authelia/authelia/v4/internal/middlewares.(*AutheliaCtx).GetSession(0xc0001d2900, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
github.com/authelia/authelia/v4/internal/middlewares/authelia_context.go:161 +0x90
github.com/authelia/authelia/v4/internal/handlers.verifyAuth(0xc0001d2900, 0xc00013a1b0, 0x0, 0x0, 0xc00045b058, 0x1, 0x1, 0x154991776108, 0x30, 0xc000342ea0, ...)
github.com/authelia/authelia/v4/internal/handlers/handler_verify.go:423 +0x18f
github.com/authelia/authelia/v4/internal/handlers.VerifyGet.func1(0xc0001d2900)
github.com/authelia/authelia/v4/internal/handlers/handler_verify.go:475 +0x327
github.com/authelia/authelia/v4/internal/middlewares.AutheliaMiddleware.func1.1(0xc0004b8580)
github.com/authelia/authelia/v4/internal/middlewares/authelia_context.go:51 +0x166
github.com/fasthttp/router.(*Router).Handler(0xc000263c00, 0xc0004b8580)
github.com/fasthttp/router@v1.4.1/router.go:414 +0x7c3
github.com/authelia/authelia/v4/internal/middlewares.LogRequestMiddleware.func1(0xc0004b8580)
github.com/authelia/authelia/v4/internal/middlewares/log_request.go:14 +0xe2
github.com/authelia/authelia/v4/internal/middlewares.StripPathMiddleware.func1(0xc0004b8580)
github.com/authelia/authelia/v4/internal/middlewares/strip_path.go:21 +0xd0
github.com/valyala/fasthttp.(*Server).serveConn(0xc000124900, 0x2a241e0, 0xc0000a8050, 0x0, 0x0)
github.com/valyala/fasthttp@v1.28.0/server.go:2239 +0x14a2
github.com/valyala/fasthttp.(*workerPool).workerFunc(0xc0004f8960, 0xc00025a280)
github.com/valyala/fasthttp@v1.28.0/workerpool.go:223 +0xba
github.com/valyala/fasthttp.(*workerPool).getCh.func1(0xc0004f8960, 0xc00025a280, 0x1112180, 0xc00025a280)
github.com/valyala/fasthttp@v1.28.0/workerpool.go:195 +0x35
created by github.com/valyala/fasthttp.(*workerPool).getCh
github.com/valyala/fasthttp@v1.28.0/workerpool.go:194 +0x11f
```

```
$ docker image inspect 564d45bfafad
[
    {
        "Id": "sha256:564d45bfafad973de93720ac066a1ad88d5e2d8b9173ddb452cc0dff6339ba10",
        "RepoTags": [
            "authelia/authelia:latest"
        ],
        "RepoDigests": [
            "authelia/authelia@sha256:956a7577adeb3d430bb1fc00b08baa8aecddec3949dd767363a43fcc6465d34a"
        ],
        "Parent": "",
        "Comment": "",
        "Created": "2021-08-11T01:51:12.03904675Z",
```

**Comment by james-d-elliott** [comment]
Can you verify the version that you're running?

**Comment by laur89** [comment]
`v4.30.4`
Yes, am using Redis. So far only seen this once.

**Comment by james-d-elliott** [comment]
@laur89 Could you confirm you're actually using redis and not a "redis compatible" alternative?

**Comment by laur89** [comment]
Running redis v6.2.5, tracking [redislabs/rejson](https://hub.docker.com/r/redislabs/rejson/) latest image.
Full authelia conf available [here](https://gist.github.com/laur89/ed9d1d22a265070fa1d14e5d93635c51)

**Comment by james-d-elliott** [comment]
We don't implement the rejson client at all. This is likely the cause of your issue. We only implement the standard redis client.

**Comment by laur89** [comment]
Ah that makes sense, thank you.
Any idea why the issue occurs only every now and then though?

**Comment by james-d-elliott** [comment]
> 
> 
> Ah that makes sense, thank you.
> Any idea why the issue occurs only every now and then though?

I'd only be guessing, since I know basically zero about the rejson implementation. My guess is some conversion that rejson does to non-json data is inconsistently applied to our use of redis storage... however I can't think of what since we just store an encrypted blob.

**Comment by laur89** [comment]
Right. I'll migrate to vanilla redis and monitor authelia.

**Comment by laur89** [comment]
Reproduced with [vanilla redis](https://registry.hub.docker.com/_/redis/) v6.2.5.

I suspect some race condition problem. Looking at the logs, we see the requests from drone service come in bursts of 10. Issue always occurs when going through a batch of 10, and it happens somewhere in the middle of it.

**Comment by laur89** [comment]
Authelia docker cmd:
```
docker create --name='authelia' --net='main' -e TZ="Europe/Paris" -e HOST_OS="Unraid" \
  -e 'PUID'='99' -e 'PGID'='100' -p '9092:9091/tcp' -v '/host/Authelia':'/config':'rw' \
  --log-opt max-size=50m --log-opt max-file=1 'authelia/authelia' 

```

Drone docker cmd:
```
docker create --name='drone' --net='main' -e TZ="Europe/Paris" -e HOST_OS="Unraid" \
  -e 'DRONE_GITEA_CLIENT_ID'='redacted' -e 'DRONE_GITEA_CLIENT_SECRET'='redacted' \
  -e 'DRONE_GITEA_SERVER'='https://git.mydomain.com' -e 'DRONE_RPC_SECRET'='redacted' \
  -e 'DRONE_SERVER_HOST'='drone.mydomain.com' -e 'DRONE_SERVER_PROTO'='https' \
  -e 'DRONE_DATABASE_DRIVER'='mysql' -e 'DRONE_DATABASE_DATASOURCE'='drone:password-redacted@tcp(db:3306)/drone?parseTime=true' \
  -e 'DRONE_DATABASE_SECRET'='redacted' -e 'DRONE_TLS_AUTOCERT'='false' \
  -e 'DRONE_USER_CREATE'='username:myuser,admin:true' -e 'PUID'='1000' -e 'PGID'='100' \
  -e 'DRONE_EXEC_DISABLED'='true' -e 'DRONE_RUNNER_CAPACITY'='10' \
  -e 'DRONE_GITHUB_CLIENT_ID'='redacted' -e 'DRONE_GITHUB_CLIENT_SECRET'='redacted' \
  -p '8086:8080/tcp' -v '/host/drone-ci/':'/mnt/config':'rw' \
  -v '/var/run/docker.sock':'/run/docker.sock':'rw' --log-opt max-size=50m --log-opt max-file=1 'nephatrine/drone-ci' 

```

**Comment by james-d-elliott** [comment]
> 
> 
> Reproduced with [vanilla redis](https://registry.hub.docker.com/_/redis/) v6.2.5.
> 
> I suspect some race condition problem. Looking at the logs, we see the requests from drone service come in bursts of 10. Issue always occurs when going through a batch of 10, and it happens somewhere in the middle of it.

That's really strange. What's generating the requests to the drone backend that cause this?

**Comment by laur89** [comment]
Assuming it's some idle-worker, as no repositories has been even enabled yet.

**Comment by james-d-elliott** [comment]
> 
> 
> Assuming it's some idle-worker, as no repositories has been even enabled yet.

I'm going to attempt a change which may fix it and narrow down the source of the issue. However the fix is unlikely to be a permanent solution until we get fasthttp to take a look.

**Comment by james-d-elliott** [comment]
You are welcome to test the authelia/authelia:fix-fasthttp-session-id image.

**Comment by james-d-elliott** [comment]
@laur89 can you confirm you've used it and the problem is vanquished by the change?

**Comment by laur89** [comment]
I'm using it as of two days ago. Need at least a week, better 2, to confirm the issue's gone.

**Comment by nightah** [comment]
Closed by #2471.

This will be in the next release.
